### DISCUSSION TITLE
Problem with gpiozero in systemd service unit

### DISCUSSION THREAD

**Discussion by gldrplt** [gpiozero/gpiozero#1153] [Q&A] [answered]
I am having a problem with a simple Python program that monitors a gpio pin.
If I run the program from a terminal, it works fine.
If I run the program from a systemd service unit it doesn't work.

I am running a raspberry pi 5 with Bookworm
No virtualenv
here is the output to stderr:

```
2024-06-08 08:28:36 xCreatePipe: Can't set permissions (436) for //.lgd-nfy0, No such file or directory
/usr/lib/python3/dist-packages/gpiozero/devices.py:295: PinFactoryFallback: Falling back from lgpio: [Errno 2] No such file or directory: '.lgd-nfy-3'
  warnings.warn(
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/gpiozero/pins/pi.py", line 408, in pin
    pin = self.pins[info]
          ~~~~~~~~~^^^^^^
KeyError: PinInfo(number=40, name='GPIO21', names=frozenset({'BOARD40', 21, 'WPI29', 'GPIO21', 'BCM21', '21', 'J8:40'}), pull='', row=20, col=2, interfaces=frozenset({'', 'pcm', 'spi', 'dpi', 'gpio'}))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/paul/Scripts/shutdown_pi_tkinter.py", line 69, in <module>
    btn = Button(21)
          ^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/gpiozero/devices.py", line 103, in __call__
    self = super().__call__(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/gpiozero/input_devices.py", line 412, in __init__
    super().__init__(
  File "/usr/lib/python3/dist-packages/gpiozero/mixins.py", line 417, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/gpiozero/input_devices.py", line 162, in __init__
    super().__init__(
  File "/usr/lib/python3/dist-packages/gpiozero/mixins.py", line 243, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/gpiozero/input_devices.py", line 79, in __init__
    super().__init__(pin, pin_factory=pin_factory)
  File "/usr/lib/python3/dist-packages/gpiozero/devices.py", line 549, in __init__
    pin = self.pin_factory.pin(pin)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/gpiozero/pins/pi.py", line 410, in pin
    pin = self.pin_class(self, info)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/gpiozero/pins/rpigpio.py", line 101, in __init__
    GPIO.setup(self._number, GPIO.IN, self.GPIO_PULL_UPS[self._pull])
RuntimeError: Cannot determine SOC peripheral base address
^^^^^^^^^^^^^^^^^
  File "/home/paul/.virtualenvs/myGarage/lib/python3.11/site-packages/gpiozero/pins/pi.py", line 413, in pin
    pin = self.pin_class(self, info)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/paul/.virtualenvs/myGarage/lib/python3.11/site-packages/gpiozero/pins/rpigpio.py", line 101, in __init__
    GPIO.setup(self._number, GPIO.IN, self.GPIO_PULL_UPS[self._pull])
RuntimeError: Cannot determine SOC peripheral base address
paul@desktop-pi-5:~ $ 
```

I have other python programs using gpiozero that work fine when launched by a systemd service.

Any suggestions?

**Comment by bennuttall** [comment]
It's failing to use lgpio as the pin factory and falling back to RPi.GPIO but that's also failing (because it's a Pi 5).

This is the only part of the error that matters:

```
2024-06-08 08:28:36 xCreatePipe: Can't set permissions (436) for //.lgd-nfy0, No such file or directory
/usr/lib/python3/dist-packages/gpiozero/devices.py:295: PinFactoryFallback: Falling back from lgpio: [Errno 2] No such file or directory: '.lgd-nfy-3'
```

I assume your systemd service needs configuring to allow GPIO access. Is the service running as the `paul` user or another user?

It would be a useful addition to the docs - how to configure a systemd service.

**Reply by gldrplt** [reply]
The systemd service is running as the paul user. 
How do I configure the systemd service to allow GPIO?

 here is the .service file
```
[Unit]
Description=My Shutdown Service - Watch GPIO21
After=graphical.target
Wants=graphical.target
DefaultDependencies=no

[Service]
Type=simple
User=paul
Group=paul
Environment="DISPLAY=:0.0"
Environment="XAUTHORITY=/home/paul/.Xauthority"
Restart=always
RemainAfterExit=true
ExecStart=/usr/bin/python3 /home/paul/Scripts/shutdown_pi_tkinter.py

[Install]
WantedBy=graphical.target
```

**Reply by bennuttall** [reply]
I'm not sure. @waveform80 @lurch any ideas?

@gldrplt can you please use three backticks to format the code block properly

**Reply by gldrplt** [reply]
bennuttall,
I'm not sure what you mean by three backsticks.  When I press the code button in the editor, it places `` in the message. Then I placed the code between the backticks.

below  I typed three backticks and typed some code
``` 
code line 1
code line 2
```

is that better?
Thanks

**Reply by bennuttall** [reply]
Your messages were surrounded by a single backtick, which is for `inline code` in a single line. It doesn't work for a multi-line code block. I edited your messages to change it, as it's hard to read otherwise.

If you select multiple lines and then click the `< >` button it does the right thing.

**Reply by lurch** [reply]
I believe you're probably running into this problem: https://github.com/joan2937/lg/issues/12

**Reply by gldrplt** [reply, marked as answer]
Lurch,
Thanks - I modified the systemd service to set a working directory and now it works as expected.
Maybe the docs could be updated to talk about systemd service units and the need to set a working directory that is unique

The default working directory of a systemd service is /.  That is what was causing the problem.  Two solutions:
1. Put WorkingDirectory=<directory>  in the service file. OR
2. use os.chdir(<directory>) in the python program before any calls to gpizero

Now the .lgd-nfy0 file gets created in <directory> instead of /

Both solutions worked for me

Thanks for your help

**Comment by gldrplt** [comment]
Got it  - thanks

**Comment by eisenhowerj** [comment]
In case you need WorkingDirectory for other purposes, you can set the environment variable:

```
[Service]
...
Environment="LG_WD=/tmp"
...
```
